src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,363 @@
+
+resolveEntries :: Mappend -> PrefixEntry h -> PrefixEntry h -> PrefixEntry h
+resolveEntries mrg e1 e2 =
+    assert (prefixEntryInvariant e1) $
+    assert (prefixEntryInvariant e2) $
+    case e1 of
+      FullEntry (Mupdate v1) ->
+        -- for monoidal resolution, we need the full second value to be present
+        FullEntry (combine mrg (Mupdate v1) (toFullEntry e2))
+      _ ->
+        -- otherwise, just take the first one (OverflowEntry can't be Mupdate!)
+        e1
+
+-- | This function should be called when discarding a 'Merge' before it
+-- was done (i.e. returned 'MergeComplete'). This removes the incomplete files
+-- created for the new run so far and avoids leaking file handles.
+--
+-- Once it has been called, do not use the 'Merge' any more!
+close ::
+     HasFS IO h
+  -> Merge (FS.Handle h)
+  -> IO ()
+close fs Merge {..} = do
+    Builder.close fs mergeBuilder
+    closeReaders fs mergeReaders
+
+{-------------------------------------------------------------------------------
+  Read
+-------------------------------------------------------------------------------}
+
+-- | Abstraction for the collection of 'RunReader', yielding elements in order.
+-- More precisely, that means first ordered by their key, then by the input
+-- run they came from. This is important for resolving multiple entries with the
+-- same key into one.
+--
+-- Construct with 'newReaders', then keep calling 'getFromReaders'.
+-- If aborting early, remember to call 'closeReaders'!
+data Readers fhandle = Readers {
+      readersHeap     :: !(Heap.MutableHeap RealWorld (ReadCtx fhandle))
+      -- | Since there is always one reader outside of the heap, we need to
+      -- store it separately.
+    , readersLastRead :: !(IORef (RunReader fhandle, ReaderNumber))
+    }
+
+newtype ReaderNumber = ReaderNumber Int
+  deriving (Eq, Ord)
+
+-- | Each heap element needs some more context than just the reader.
+-- E.g. the 'Eq' instance we need to be able to access the first key to be read
+-- in a pure way.
+--
+-- TODO(optimisation): We allocate this record for each k/op. This might be
+-- avoidable, see ideas below.
+data ReadCtx fhandle = ReadCtx {
+      -- We could avoid this using a more specialised mutable heap with separate
+      -- arrays for keys and values (or even each of their components).
+      -- Using an 'STRef' could avoid reallocating the record for every entry,
+      -- but that might not be straightforward to integrate with the heap.
+      readCtxHeadKey   :: !SerialisedKey
+    , readCtxHeadEntry :: !(PrefixEntry fhandle)
+      -- We could get rid of this by making 'LoserTree' stable (for which there
+      -- is a prototype already).
+      -- Alternatively, if we decide to have an invariant that the number in
+      -- 'RunFsPaths' is always higher for newer runs, then we could use that
+      -- in the 'Ord' instance.
+    , readCtxNumber    :: !ReaderNumber
+    , readCtxReader    :: !(RunReader fhandle)
+    }
```

Comment:
Using mutable structures to prevent re-allocating should be useful. There are probably two dimensions to this:

* Have a single mutable `ReadCtx` per run, which gets re-used
* Put only data that is important for comparing elements (i.e., the key and the run number) in the `ReadCtx`, and instead use mutable structures outside of the heap to communicate data back and forth. For example, we would have something like an `MVector s (RunRead fhandle)` and `MVector s (PrefixEntry fhandle)`. 

Both options would probably require changes to the heap code to be able to do ST-like computations when inserting/extracting on the heap. I haven't worked out the details of something like that would be possible, but interesting to think about after this PR